genie-agent/src/main/java/com/netflix/genie/agent/execution/services/impl/UploadServiceImpl.java

```diff
@@ -0,0 +1,145 @@
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services.impl;
+
+import com.netflix.genie.agent.execution.exceptions.UploadException;
+import com.netflix.genie.agent.execution.services.UploadService;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.core.io.WritableResource;
+import org.springframework.stereotype.Component;
+
+import javax.validation.constraints.NotNull;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.net.URI;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipOutputStream;
+
+/**
+ * Implementation of UploadService.
+ *
+ * @author standon
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class UploadServiceImpl implements UploadService {
+    private final ResourceLoader resourceLoader;
+
+    UploadServiceImpl(final ResourceLoader resourceLoader) {
+        this.resourceLoader = resourceLoader;
+    }
+
+    /**
+     * Compress the file and upload to a target location. For a directory
+     * all its contents are recursively compressed and uploaded to the target
+     * location
+     *
+     * @param file              file/directory to upload.
+     * @param targetLocationURI URI representing the target location for the file
+     * @throws UploadException if the upload fails
+     */
+
+    @Override
+    public void compressAndUpload(
+        @NotNull final File file,
+        @NotNull final URI targetLocationURI
+    ) throws UploadException {
+
+        final String uriString = targetLocationURI.toASCIIString();
+
+        log.debug("Uploading to target location: {}", uriString);
+
+        final WritableResource writableResource = (WritableResource) resourceLoader.getResource(uriString);
+
+        try (ZipOutputStream outputStream = new ZipOutputStream(writableResource.getOutputStream())) {
+            zipDirectory(file, file.getName(), outputStream);
+        } catch (Exception e) {
+            log.info("Error uploading to target location: {} ", uriString);
+            throw new UploadException("Error uploading file: " + file.getName(), e);
+        }
+    }
+
+    /**
+     * Zip a directory recursively.
+     *
+     * @param fileToZip         file/dir to zip
+     * @param nameInArchive     name for the file entry in archive
+     * @param zipOutputStream   destination output stream
+     * @throws IOException      in case of a problem
+     */
+    private void zipDirectory(
+        @NotNull final File fileToZip,
+        @NotNull final String nameInArchive,
```

Comment:
`nameInArchive` is the relative path w.r.t. the root of the job directory, no?
Why not name it as such?

Reply:
Sure

Reply:
On giving it a thought, I think this method signature can overall be improved. `nameInArchive/relativePathToRoot` is derivable. The method should be something like `zipRecursively(File fileToZip, File root, Outputstream os)`. I will work on making this cleaner